### ISSUE TITLE
CI4: Stale data in appconfig update of cache

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#3693]
Currently, when the an app_config key/value pair is updated or inserted into the database, the cash gets invalidated then updated.  There are two problems:
1- Calling the update_settings() function in the save() function causes a lot of calls to invalidating and updating the cache.  Perhaps there is a better place to call this.
2- calling update_settings() invalidates and updates the cache, so any future instances of config('OSPOS') will be current, but existing instances of the config are stale until they are regenerated.  This could cause data to be updated in the database and cache but not in the in-memory values.

For 1, We could call update_settings() in the batch_save() function after the foreach loop.  This will cut down on all but three calls to save() directly.  Then we could call update_settings() after those three calls.  The potential risk is that if we add more calls to save() directly and forget to call update_settings() we will end up with stale data in the cache.

As for 2, we need to create a new instance of config('OSPOS') after update_settings is run. The problem with this method is that it's decentralized.  Perhaps the best option here is to create a wrapper function and pass the config() instance by reference, then call both the save() function, update_settings() and refresh the config instance together.  We may need two functions or to just call batch_save or save from the wrapper based on whether the passed parameters are a single row or multiple.

**Comment by jekkos** [comment]
The more pragmatic solution might be to invalidate the whole cache once something is updated through the model. The crazy thing is that in fact the cache is by default file backed, so instead of reading the values from the db, we read them from a file, which might not be all that faster in the end?

There is no easy in memory solution unless if we enable redis or memcached. But doing this only for this purpose feels a bit overkill.

**Comment by objecttothis** [comment]
Yes, the default is file, but I believe I changed that to be session based since we use sessions in CI3

**Comment by objecttothis** [comment]
Also, we are loading the table all at once currently (if the cache is not populated), storing it in memory cache (after pulling it) and in the global array which all gets repeated if a user makes a change to one of the elements of the array.  As I said in the other PR, I don't think we should get rid of it just for the sake of getting rid of it (since it's working... I think), but only if it serves a purpose, such as making the code empty value resilient without having to add in checks for empty values.

**Comment by jekkos** [comment]
> Yes, the default is file, but I believe I changed that to be session based since we use sessions in CI3

I do believe the appconfig table was fully reloaded in memory at the start of every request in CI3.. I don''t think it was linked to the session at that point. (https://github.com/opensourcepos/opensourcepos/blob/master/application/hooks/load_config.php) . This hook is called pre controller constructor.

That has changed now as we keep the table contents as a json array in a cache that is backed by a file, with ttl of 5 mins. I'm ok to keep it if it works and if we don't need to do any cache eviction once we update the config and refresh it from the database. I would certainly keep the class but then change the implementation a bit so that it can deal with array dereference and missing values directly.

btw @objecttothis do you live in Nagorno Karabach currently? Was just reading the news on the situation in Azerbeidzjan. Hope you're all fine.

**Comment by objecttothis** [comment]
@owlbrudder let's move the conversation here, rather than keeping it going in a merged PR.
> Can you clarify for me what you mean by the standard CI4 way of working?

https://codeigniter4.github.io/userguide/general/configuration.html#getting-a-config-property
CI4 has a means of accessing built-in configuration files and allowing the user to specify and modify custom configuration files. I had considered the idea of completely scrapping the CI3 implementation, establishing a custom config file with an associative array and populating that from the database, but decided to try to make something similar to our current implementation working since the CI4 conversion was immense as it was.

**Comment by objecttothis** [comment]
> btw @objecttothis do you live in Nagorno Karabach currently? Was just reading the news on the situation in Azerbeidzjan. Hope you're all fine.
I'm far enough away to not be in any danger. It seems as if the fighting has stopped anyway.

>That has changed now as we keep the table contents as a json array in a cache that is backed by a file, with ttl of 5 mins. I'm ok to keep it if it works and if we don't need to do any cache eviction once we update the config and refresh it from the database. I would certainly keep the class but then change the implementation a bit so that it can deal with array dereference and missing values directly.
Beyond the hope of performance increase, I can't remember now what was driving me to store the array in cache. I can't remember if not doing so was affecting accessibility or not. Feel free to tweak the current implementation as you see fit.

My suggestion is that we use the current implementation if it doesn't mean peppering the code with a bunch of checks and then after CI4 has been released, we consider simplifying it in a separate PR.  Thoughts?

**Comment by owlbrudder** [comment]
> @owlbrudder let's move the conversation here, rather than keeping it going in a merged PR.
> 
> > Can you clarify for me what you mean by the standard CI4 way of working?
> 
> https://codeigniter4.github.io/userguide/general/configuration.html#getting-a-config-property CI4 has a means of accessing built-in configuration files and allowing the user to specify and modify custom configuration files. I had considered the idea of completely scrapping the CI3 implementation, establishing a custom config file with an associative array and populating that from the database, but decided to try to make something similar to our current implementation working since the CI4 conversion was immense as it was.

Thank you for the clarification. I agree about the scale of the conversion.

**Comment by objecttothis** [comment]
> I expect 99.9% of the config[] problems will go away if we accept the proposal to write a new migration script to populate the database with all keys we dereference later. This avoids wrapping every dereference in a protective layer. Far better to focus on getting what we have working reliably than to rewrite at this point. Let rewrites become later point releases, but make the existing system work first.

Sounds good to me. If @jekkos isn't opposed, I think you can proceed. There should be an existing migration for ci4. I'd put those changes in that migration so that we don't end up with 100's of migrations.

**Comment by owlbrudder** [comment]
> > I expect 99.9% of the config[] problems will go away if we accept the proposal to write a new migration script to populate the database with all keys we dereference later. This avoids wrapping every dereference in a protective layer. Far better to focus on getting what we have working reliably than to rewrite at this point. Let rewrites become later point releases, but make the existing system work first.
> 
> Sounds good to me. If @jekkos isn't opposed, I think you can proceed. There should be an existing migration for ci4. I'd put those changes in that migration so that we don't end up with 100's of migrations.

@jekkos has already pointed me to the migration script for the existing database population, so I expect I can extend that very quickly. I will get onto it tomorrow.

**Comment by owlbrudder** [comment]
There are cases in which just having the array key in existence is not enough to avoid problems. The migration script is now creating a record for 'mailchimp_api_key' but the value is the empty string. In Controllers/Config.php, for example, we need to also test if the value of 'mailchimp_api_key' is not empty before passing it to Services::encrypter->decrypt() because that crashes if not passed a valid string. This would be overcome if we implement an array dereferencing mechanism which tests for both existence and content. In the meantime, I have to submit a revision to Controllers/Config.php to implement this test. Previously it crashed because the key did not exist; now it crashes because the value is empty. It is reasonable to assume other cases where we create an app-config table record with empty data will raise similar problems.
